dev_scripts/qa.py

```diff
@@ -0,0 +1,448 @@
+            def inner(self, *args, **kwargs):
+                # Assume that the first line is the description of the task, and
+                # decorate it with some "=" signs.
+                dec_msgs = list(msgs)
+                num_equals = (80 - 2 - len(msgs[0])) // 2 or 3
+                dec_msgs.insert(0, "")
+                dec_msgs[1] = "=" * num_equals + " " + msgs[0] + " " + "=" * num_equals
+                self.describe(*dec_msgs)
+
+                # Detect if the task should run automatically, either because the user
+                # has provided the `--try-auto` flag, or because they typed "auto" when
+                # prompted.
+                try_auto = False
+                if self.try_auto:
+                    try_auto = True
+                else:
+                    prompt = "Type 'auto' to try automatically, or ENTER to skip: "
+                    choice = self.prompt(ref=ref, prompt=prompt)
+                    try_auto = "auto" == choice
+
+                # Run the task automatically, if the user asked to, else assume that the
+                # user has run it manually.
+                if try_auto:
+                    logger.info("Trying automatically...")
+                    out = func(self, *args, **kwargs)
+                    logger.info("Completed successfully")
+                    return out
+                else:
+                    logger.debug("Assuming step %s has completed manually...", func)
+
+            return inner
+
+        return decorator
+
+    def run(self, *args):
+        """Run a command with extra debug logs."""
+        # Construct the command line representation of the provided arguments, by taking
+        # the arguments and joining them with a space character. If an argument already
+        # has a space character in it, quote it before adding it to the CLI
+        # representation.
+        cmd_str = ""
+        for arg in args:
+            if " " in arg:
+                cmd_str += " '" + arg + "'"
+            else:
+                cmd_str += " " + arg
+
+        logger.debug("Running %s", cmd_str)
+        try:
+            subprocess.run(args, check=True)
+            logger.debug("Successfully ran %s", cmd_str)
+        except subprocess.SubprocessError as e:
+            logger.debug("Failed to run %s", cmd_str)
+            raise
+
+    def try_run(self, *args):
+        """Try to run a command and return True/False, depending on the result."""
+        try:
+            self.run(*args)
+            return True
+        except subprocess.SubprocessError as e:
+            return False
+
+    def describe(self, *msgs, ref=None):
+        """Print a task's description.
+
+        Log one message per line, and optionally allow a URL as a reference.
+        """
+        for msg in msgs:
+            logger.info(msg)
+        if ref:
+            logger.info("For reference, see: %s" % ref)
+
+    def prompt(self, *msgs, ref=None, prompt=None):
```

Comment:
Interesting categorization, I agree with that. As for the two suggestions, I think that asking the user for an explanation would probably not amount to much, and will add extra complexity to the code. As for changing this to a `[Y/n]` prompt, I'd agree, since it is aligned with [this discussion](https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1049674100).

My concern is that us devs may have partially run QA on a platform, and may want a quick way to reach the point where we left off, so the `[Y/n]` prompts may prove frustrating. We'll see...